tests/chainerx_tests/unit_tests/routines_tests/test_linalg.py

```diff
@@ -93,3 +93,135 @@ def test_dot_invalid(is_module, xp, device, a_shape, b_shape, dtype):
         return xp.dot(a, b)
     else:
         return a.dot(b)
+
+
+@pytest.mark.parametrize('dtype', ['float32', 'float64'])
+class NumpyLinalgOpTest(op_utils.NumpyOpTest):
+
+    dodge_nondifferentiable = True
+
+    def setup(self, dtype):
+        device = chainerx.get_default_device()
+        if (device.backend.name == 'native'
+                and not chainerx.linalg._is_lapack_available()):
+            pytest.skip('LAPACK is not linked to ChainerX')
+        self.dtype = dtype
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    chainer.testing.product({
+        'shape': [(1, 1), (3, 3), (6, 6)],
+        'b_columns': [(), (1,), (3,), (4,)]
+    })
+))
+class TestSolve(NumpyLinalgOpTest):
+
+    def generate_inputs(self):
+        a = numpy.random.random(self.shape).astype(self.dtype)
+        b = numpy.random.random(
+            (self.shape[0], *self.b_columns)).astype(self.dtype)
+        return a, b
+
+    def forward_xp(self, inputs, xp):
+        a, b = inputs
+        out = xp.linalg.solve(a, b)
+        return out,
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    chainer.testing.product({
+        'shape': [(2, 3), (3, 2)]
+    })
+))
+class TestSolveFailing(NumpyLinalgOpTest):
+
+    forward_accept_errors = (numpy.linalg.LinAlgError,
+                             chainerx.DimensionError)
+
+    def generate_inputs(self):
+        a = numpy.random.random(self.shape).astype(self.dtype)
+        b = numpy.random.random(self.shape).astype(self.dtype)
+        return a, b
+
+    def forward_xp(self, inputs, xp):
+        a, b = inputs
+        out = xp.linalg.solve(a, b)
+        return out,
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    chainer.testing.product({
+        'shape': [(1, 1), (3, 3), (6, 6)]
+    })
+))
+class TestInverse(NumpyLinalgOpTest):
+
+    def generate_inputs(self):
+        a = numpy.random.random(self.shape).astype(self.dtype)
+        return a,
+
+    def forward_xp(self, inputs, xp):
+        a, = inputs
+        out = xp.linalg.inv(a)
+        return out,
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    chainer.testing.product({
+        'shape': [(2, 3), (3, 2)]
+    })
+))
+class TestInverseFailing(NumpyLinalgOpTest):
+
+    forward_accept_errors = (numpy.linalg.LinAlgError,
+                             chainerx.DimensionError)
+
+    def generate_inputs(self):
+        a = numpy.random.random(self.shape).astype(self.dtype)
+        return a,
+
+    def forward_xp(self, inputs, xp):
+        a, = inputs
+        out = xp.linalg.inv(a)
+        return out,
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize_pytest('shape', [(3, 3)])
+@chainer.testing.parameterize_pytest('dtype', ['float16'])
+class TestInverseDtypeFailing(op_utils.NumpyOpTest):
```

Comment:
Should derive `NumpyLinalgOpTest` (test must be conditionally skipped depending on LAPACK availability)?

Reply:
Yes, it should derive `NumpyLinalgOpTest`.
I included dtype parametrizations in the `NumpyLinalgOpTest` class and
`@chainer.testing.parameterize_pytest('dtype', ['float16'])`
does not override the baseclass' dtype, therefore here I used `op_utils.NumpyOpTest` but I forgot checking against not-installed-lapack build.
I guess I have to add back fp32, fp64 parametrizations explicitly to all other tests.

Reply:
Yes, I think that's fine. It makes sense to specify supported dtypes per routine.